lib/cdo/aws/s3.rb

```diff
@@ -1,29 +1,55 @@
-require 'aws/s3'
 require 'aws-sdk'
 
 module AWS
   module S3
-    # Connects to the official AWS SDK for Ruby v2
+
+    # Region for storing S3 buckets. TODO: Move this CDO configuration.
+    S3_REGION = 'us-east-1'
+
+    # An exception class used to wrap the underlying Amazon NoSuchKey exception.
+    class NoSuchKey < Exception
+      def initialize(message = nil)
+        super(message)
+      end
+    end
+
+    # Creates an S3 client using the the official AWS SDK for Ruby v2 and
+    # the credentials specified in the CDO config.
+    # @return [Aws::S3::Client]
     def self.connect_v2!
-      s3_params = {access_key_id: CDO.s3_access_key_id, secret_access_key: CDO.s3_secret_access_key, region: 'us-east-1'}
+      s3_params = {access_key_id: CDO.s3_access_key_id,
+                   secret_access_key: CDO.s3_secret_access_key,
+                   region: S3_REGION}
       Aws::S3::Client.new(s3_params)
     end
 
-    def self.connect!()
-      AWS::S3::Base.establish_connection! access_key_id: CDO.s3_access_key_id, secret_access_key: CDO.s3_secret_access_key
+    # A simpler name for connect_v2!
+    def self.create_client
+      self.connect_v2!
     end
 
-    def self.download_from_bucket(bucket, name, options={})
-      connect!
-      AWS::S3::S3Object.find(name, bucket).value
+    # Returns the value of the specified S3 key in bucket.
+    # @param [String] bucket
+    # @param [String] key
+    # @return [String]
+    def self.download_from_bucket(bucket, key, options={})
+      create_client.get_object(bucket: bucket, key: key).body.string
```

Comment:
Fixed and added tests to verify that arbitrary byte sequences round-trip correctly.

(Both the .string and the .read method were defaulting to utf-8 encoding for the bytes read from the bucket, so it was necessary to force the encoding to BINARY.)